### PR TITLE
blend functions

### PR DESCRIPTION
**Pull request by mkborregaard**
This is just to signal that I've started porting the last utility functions

### CODE REVIEW

**Comment by mkborregaard** [comment]
Can I get an opinion here @tpoisot @gottacatchenall ?
It doesn't make much sense to port the clumsy numpy code directly for these functions, so I've reimplemented. But there is one section in the original code:
https://github.com/tretherington/nlmpy/blob/master/nlmpy/nlmpy.py#L271-L273

It looks like they're using the `meanCluster` function to replace all cluster values with mean-per-cluster values - which makes sense - but then they "blend" (average) those values together with the actual values for the original clusters. Does that make sense at all? What's going on?

**Comment by tpoisot** [comment]
It looks like an average but what if the number of additions is not the same? Should we tag the original authors or look at the Cpp version? I think @gottacatchenall started reading it.

**Comment by mkborregaard** [comment]
do we know where the cpp version lives?

**Comment by tpoisot** [comment]
https://github.com/ropensci/NLMR - they added a fair number of algorithms as well, I would suggest we tag a release at feature parity with NLMpy, then work on NLMR

**Comment by mkborregaard** [comment]
That would require re-licensing to GPL-3 though.

**Comment by mkborregaard** [comment]
The blending functions don't appear to exist in NLMR

**Comment by rafaqz** [comment]
We can probably just find the specific algorithms in other MIT licensed packages, e.g. fractal brownian motion is available here:  https://github.com/crflynn/fbm

**Comment by mkborregaard** [comment]
When reviewing, could you check that these functions do what the python functions do? I have not ported directly but rewritten in Julia, and I haven't checked (I've never used python on this machine, though I guess I do have it).

File: src/classify.jl [hunk 1]
```diff
@@ -16,3 +16,44 @@ function classify!(array, weights, mask = nothing)
...
+end
+
+"""
+    blendClusterArray(primary, arrays[, scaling])
+
+Blend a primary cluster NLM with other arrays in which the mean value per 
+cluster is weighted by scaling factors.
+"""
+function blendClusterArray(primary, arrays, scaling = ones(length(arrays)))
+    ret = sum(_clusterMean.(Ref(primary), arrays) .* scaling)
```
Comment (mkborregaard) [inline src/classify.jl]: I guess I could have just used the `blend` function too...

**Comment by tpoisot** [comment]
> That would require re-licensing to GPL-3 though.

Let's abandon this idea immediately then...

File: src/classify.jl (hunk 1 above)
Comment (tpoisot) [reply src/classify.jl]: I think so, the only difference is calling it with a matrix and a vector of matrices vs. a vector of matrices alone.

File: src/classify.jl [hunk 2]
```diff
@@ -16,3 +16,44 @@ function classify!(array, weights, mask = nothing)
     end
     array
 end
+
+"""
+    blend(arrays[, scaling])
+
+Blend arrays weighted by scaling factors.
+"""
+function blend(arrays, scaling = ones(length(arrays)))
```
Comment (tpoisot) [inline src/classify.jl]: test for `length(arrays) == length(scaling)` sounds like a good idea here?

File: src/classify.jl [hunk 3]
```diff
@@ -16,3 +16,44 @@ function classify!(array, weights, mask = nothing)
...
+            clusters[temp] = clustersum[temp] = 0.0
+        end
+        clusters[temp] += 1.0
+        clustersum[temp] += array[ind]          
+    end
+    for cl in keys(clusters)
+        clustersum[cl] /= clusters[cl]
+    end
+    clustersum[NaN] = NaN
+    _rescale!(get.(Ref(clustersum), clusterArray, NaN))
```
Comment (tpoisot) [inline src/classify.jl:34]: what is the `_rescale!` `rescale!` difference?

Comment (mkborregaard) [reply src/classify.jl:34]: there is no rescale! function

File: src/classify.jl (hunk 2 above)
Comment (mkborregaard) [reply src/classify.jl]: I see the point here, but is that error message really that more helpful?
```julia
julia> scaling = 1:4;
julia> arrays = [rand(3) for i in 1:3];

julia> @assert length(arrays) == length(scaling)
ERROR: AssertionError: length(arrays) == length(scaling)
Stacktrace:
 [1] top-level scope
   @ REPL[4]:1

julia> arrays .* scaling
ERROR: DimensionMismatch("arrays could not be broadcast to a common size; got a dimension with lengths 3 and 4")
Stacktrace:
```

File: src/classify.jl (hunk 3 above)
Comment (mkborregaard) [reply src/classify.jl:34]: I updated this now

File: src/classify.jl (hunk 1 above)
Comment (mkborregaard) [reply src/classify.jl]: so would this be preferable?
```julia
blendClusterArray(primary, arrays, scaling = ones(length(arrays))) = 
    blend([_clusterMean.(Ref(primary), [primary; arrays], scaling)
```
?

Comment (mkborregaard) [reply src/classify.jl]: I might have a preference for the original

File: src/classify.jl (hunk 2 above)
Comment (tpoisot) [reply src/classify.jl]: I was thinking more along the lines of 

~~~ julia
if length(scaling) != length(arrays)
    throw(DimensionMismatch("The array of landscapes (n = $(length(arrays))) and scaling (n = $(length(scaling)) must have the same length")
end
~~~

which is more informative than both.

File: src/classify.jl (hunk 1 above)
Comment (tpoisot) [reply src/classify.jl]: my point was to maybe call *both* functions `blend`?

Comment (mkborregaard) [reply src/classify.jl]: ah, I see! gotcha

**Comment by mkborregaard** [comment]
Alright @tpoisot this should be ready to merge now

**Comment by mkborregaard** [comment]
... except I guess I should put an example in the docs

**Comment by tpoisot** [comment]
> ... except I guess I should put an example in the docs

let's do a round of doc update when everything is in

**Pull request merged**